File: src/ipc/messages/types/structs/PathComponent.ts
import * as ldm from '../../leapdatamodel';
import * as valuetype from '../../valuetype';
import type * as types from '..';
import { StructBase as _StructBase } from '../../StructBase';
import { LeapError as _LeapError } from '../../LeapError';
import { ValueBase as _ValueBase } from '../../ValueBase';

export class PathComponent extends _StructBase {
    static _type: string = 'path-component';
    static _leapStruct: valuetype.LeapStruct = new valuetype.LeapStruct("path-component", [], [new valuetype.Property("path", new valuetype.ValueType("str", [])), new valuetype.Property("name", new valuetype.ValueType("str", []))]);
    path: string;
    name: string;

    constructor(path: string, name: string, ) {
        super();
        this.path = path;
        this.name = name;
    }

    _verifyAsTypeOrRaise(path: string[], valueType: valuetype.ValueType): void {
        if (valueType.name == PathComponent._type) {
            const structType = PathComponent._leapStruct.applyArgs(valueType.args);
            _ValueBase._verifyValueOrRaise(path.concat(['path']), this.path, structType.props[0].propType);
            _ValueBase._verifyValueOrRaise(path.concat(['name']), this.name, structType.props[1].propType);
        } else {
            throw _ValueBase._pathError(path, `expecting ${PathComponent._type}`);
        }
    }

    _toLdm(valueType: valuetype.ValueType): ldm.Value {
        let value = new Map<string, ldm.Value>();
        const structType = PathComponent._leapStruct.applyArgs(valueType.args);
        value.set("path", _ValueBase._valueToLdm(this.path, structType.props[0].propType));
        value.set("name", _ValueBase._valueToLdm(this.name, structType.props[1].propType));
        return new ldm.StructValue(value, valueType);
    }
}

File: src-tauri/src/ipc/messages/types/structs/path_component.rs
use std::collections::HashMap as _HashMap;
#[allow(unused_imports)]
use super::super::super::types;
use super::super::super::value_base::ValueBase as _ValueBase;
use super::super::super::struct_base::StructBase as _StructBase;
use super::super::super::leap_data_model as ldm;
use super::super::super::value_type::ValueType as _ValueType;

#[derive(Clone, Debug)]
pub struct PathComponent {
    pub path: String,
    pub name: String,
}

impl PathComponent {
    pub fn new(path: String, name: String, ) -> Self {
        Self {
            path,
            name,
        }
    }
}

impl _ValueBase for PathComponent {
    const _TYPE: &'static str = "path-component";

    fn _applied_args() -> Vec<_ValueType> {
        vec![
        ]
    }

    fn _to_ldm(&self) -> ldm::Value {
        ldm::Value::Struct(self._to_ldm_struct())
    }
}

impl _StructBase for PathComponent {
    fn _to_ldm_struct(&self) -> ldm::StructValue {
        let mut value = _HashMap::new();
        value.insert("path".to_owned(), self.path._to_ldm());
        value.insert("name".to_owned(), self.name._to_ldm());
        ldm::StructValue{
            value,
            value_type: _ValueType::new(Self::_TYPE.to_owned(), Self::_applied_args()),
        }
    }
}